cmd/promtool/backfill.go

```diff
@@ -106,15 +115,22 @@ func createBlocks(input []byte, mint, maxt int64, maxSamplesInAppender int, outp
 					continue
 				}
 
-				l := labels.Labels{}
-				p.Metric(&l)
 				_, ts, v := p.Series()
 				if ts == nil {
+					l := labels.Labels{}
+					p.Metric(&l)
 					return errors.Errorf("expected timestamp for series %v, got none", l)
 				}
-				if *ts < t || *ts >= tsUpper {
+				if *ts < t {
 					continue
 				}
+				if *ts >= tsUpper {
+					nextSampleTs = *ts
+					break
```

Comment:
Do we have a guarantee that the samples are always ordered by timestamp? I had the impression that the ordering is only guaranteed per metric.

Reply:
If the ordering is guaranteed and we can break safely, then we would not need to read the complete file each time we want a block.

Reply:
We could just parse once.

Reply:
Of course. Damn you OpenMetrics! So what I saw as a needless complication was actually necessary. Will delete this branch quickly to hide my embarrassment.